ui/text_displayer.js

```diff
@@ -168,6 +168,133 @@ shaka.ui.TextDisplayer = class {
     }
   }
 
+  /**
+   * Returns a safe collection of html elements for display as a single cue.
+   * @param {string} cueBody
+   * @return {Array<HTMLElement>}
+   * @private
+   */
+  buildCueElements_(cueBody) {
+    const tokenize = (str) => {
+      if (!str) {
+        return [];
+      }
+
+      const tokens = [];
+
+      let currentValue = '';
+      for (let i = 0; i < str.length; i++) {
+        if (str[i] === '<' && str[i+1] !== ' ') {
+          if (currentValue) {
+            // Flush the current value as we enter tag detection.
+            tokens.push({type: 'text', value: currentValue});
+          }
+          currentValue = '<';
+
+          // Seek to end of tag.
+          i++;
+          let wellTerminated = false;
+          for (;i < str.length; i++) {
+            currentValue += str[i];
+            if (str[i] === '>') {
+              wellTerminated = true;
+              break;
+            }
+          }
+
+          if (!wellTerminated || currentValue.length < 3) {
+            // No termination found, append literal string.
+            tokens.push({type: 'text', value: currentValue});
+          } else {
+            let tagType = 'startTag';
+            currentValue = currentValue.substr(1, currentValue.length - 2);
+            if (currentValue[0] == '/') {
+              currentValue = currentValue.substr(1); // Removes /
+              tagType = 'endTag';
+            }
+            if (currentValue) {
+              tokens.push({type: tagType, value: currentValue});
+            }
+          }
+          currentValue = '';
+        } else {
+          currentValue += str[i];
+        }
+      }
+
+      // Final flush.
+      if (currentValue) {
+        tokens.push({type: 'text', value: currentValue});
```

Comment:
I think it might be preferable, for the sake of the compiler's type checking, if you made these into private methods that have their own annotations about argument and return types.  This, for example, returns !Array<{type: string, value: string}>

Reply:
Done!